src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingStructureValidationPass.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal class FormattingStructureValidationPass : FormattingPassBase
+    {
+        private readonly ILogger _logger;
+
+        public FormattingStructureValidationPass(
+            RazorDocumentMappingService documentMappingService,
+            FilePathNormalizer filePathNormalizer,
+            ILanguageServer server,
+            ILoggerFactory loggerFactory)
+            : base(documentMappingService, filePathNormalizer, server)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<CSharpOnTypeFormattingPass>();
+        }
+
+        // We want this to run at the very end.
+        public override int Order => DefaultOrder + 1000;
+
+        public override FormattingResult Execute(FormattingContext context, FormattingResult result)
+        {
+            Debug.Assert(result.Kind == RazorLanguageKind.Razor, "This method shouldn't be called for projected document edits.");
+
+            if (result.Edits.Length == 0)
+            {
+                // No op.
+                return result;
+            }
+
+            if (FormatsOutsidePureCSharpBlocks(context, result))
+            {
+                return new FormattingResult(Array.Empty<TextEdit>());
+            }
+
+            return result;
+        }
+
+        private bool FormatsOutsidePureCSharpBlocks(FormattingContext context, FormattingResult result)
+        {
+            if (!context.IsFormatOnType)
```

Comment:
Should  happen in the caller right? Feels odd to have  in a `FormatsOutsidePureCSharpBlocks` method.